Tests/Client/Custom Assertions/AssertNetworkRequest.swift

```diff
@@ -0,0 +1,150 @@
+///
+func AssertNetworkRequest(
+    method: Endpoint.Method,
+    path: String,
+    headers: [String: String]?,
+    queryParameters: [String: String]?,
+    body: [String: Any]?,
+    timeout: TimeInterval = 0.5,
+    file: StaticString = #file,
+    line: UInt = #line
+) {
+    guard let request = RequestRecorderURLProtocol.getLatestRequestSynchronously(timeout: timeout) else {
+        XCTFail("Waiting for request timed out. No request was made.", file: file, line: line)
+        return
+    }
+
+    var errorMessage = ""
+    defer {
+        if !errorMessage.isEmpty {
+            XCTFail("AssertNetworkRequest failed:" + errorMessage, file: file, line: line)
+        }
+    }
+
+    // Check method
+    if method.rawValue != request.httpMethod {
+        errorMessage += "\n  - Incorrect method: expected \"\(method.rawValue)\" got \"\(request.httpMethod ?? "_")\""
+    }
+
+    guard let url = request.url, let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
+        errorMessage += "\n  - Missing URL"
+        return
+    }
+
+    // Check path
+    if components.path != path {
+        errorMessage += "\n  - Incorrect path: expected \"\(path)\" got \"\(components.path)\""
+    }
+
+    // Check headers
+    let requestHeaders = request.allHTTPHeaderFields ?? [:]
+    headers?.forEach { (key, value) in
+        if let requestHeaderValue = requestHeaders[key] {
+            if value != requestHeaderValue {
+                errorMessage += "\n  - Incorrect header value for \"\(key)\": expected \"\(value)\" got \"\(requestHeaderValue)\""
+            }
+
+        } else {
+            errorMessage += "\n  - Missing header value for \"\(key)\""
+        }
+    }
+
+    // Check query parameters
+    let items = components.queryItems ?? []
+    queryParameters?.forEach { (key, value) in
+        if let requestValue = items[key] {
+            if value != requestValue {
+                errorMessage += "\n  - Incorrect query value for \"\(key)\": expected \"\(value)\" got \"\(requestValue)\""
+            }
+
+        } else {
+            errorMessage += "\n  - Missing query value for \"\(key)\""
+        }
+    }
+
+    // Check the request body
+    var requestBodyData: Data?
+    if let data = request.httpBody {
+        requestBodyData = data
+    }
+    if let stream = request.httpBodyStream {
+        requestBodyData = Data(reading: stream)
+    }
+
+    guard let data = requestBodyData else {
+        if body != nil {
+            errorMessage += "\n  - Missing request body"
+        }
+        return
+    }
+
+    guard let requestBodyJSON = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
+        errorMessage += "\n  - Request body is not a valid JSON"
+        return
+    }
+
+    body?.forEach { (key, value) in
+        if let requestValue = requestBodyJSON[key] {
+            if String(describing: value) != String(describing: requestValue) {
```

Comment:
I suppose the order of keys can be different. Maybe better to check this recursively.

Reply:
The order of the keys doesn't matter in this implementation.